rate-limiter/src/token_bucket.rs

```diff
@@ -1,211 +1,1209 @@
+    /// Calculates amount of time by which we should delay next call to some governed resource in order to satisfy
     /// configured rate limit.
-    pub fn rate_limit(&mut self, requested: usize, now: Instant) -> Option<Duration> {
+    pub fn rate_limit(&mut self, requested: u64) -> Option<Instant> {
         trace!(
             target: LOG_TARGET,
-            "TokenBucket called for {} of requested bytes. Internal state: {:?}.",
-            requested,
-            self
+            "TokenBucket called for {requested} of requested bytes. Internal state: {self:?}.",
         );
-        if self.requested > 0 || self.available < requested {
-            assert!(
-                now >= self.last_update,
-                "Provided value for `now` should be at least equal to `self.last_update`: now = {:#?} self.last_update = {:#?}.",
-                now,
-                self.last_update
-            );
-            if self.update_units(now) < requested {
-                self.requested = self.requested.saturating_add(requested);
-                let required_delay = self.calculate_delay();
-                return Some(required_delay);
+        let now_available = self.available().unwrap_or(0);
+        if now_available < requested {
+            self.update_tokens()
+        }
+        self.account_requested_tokens(requested);
+        let delay = self.calculate_delay();
+        trace!(
+            target: LOG_TARGET,
+            "TokenBucket calculated delay after receiving a request of {requested}: {delay:?}.",
+        );
+        delay
+    }
+}
+
+/// Determines how often instances of [SharedBandwidthManager] should check if their allocated bandwidth has changed.
+const BANDWIDTH_CHECK_INTERVAL: Duration = Duration::from_millis(250);
+
+/// Implementation of the bandwidth sharing strategy that attempts to assign equal portion of the total bandwidth to all active
+/// consumers of that bandwidth.
+pub struct SharedBandwidthManager {
+    max_rate: NonZeroRatePerSecond,
+    peers_count: Arc<AtomicU64>,
+    already_requested: Option<NonZeroRatePerSecond>,
+}
+
+impl SharedBandwidthManager {
+    /// Constructs a new instance of [SharedBandwidthManager] configured with a given rate that will be shared between all
+    /// calling consumers (clones of this instance).
+    pub fn new(max_rate: NonZeroRatePerSecond) -> Self {
+        Self {
+            max_rate,
+            peers_count: Arc::new(AtomicU64::new(0)),
+            already_requested: None,
+        }
+    }
+
+    pub fn share(&self) -> Self {
+        Self {
+            max_rate: self.max_rate,
+            peers_count: self.peers_count.clone(),
+            already_requested: None,
+        }
+    }
+
+    fn calculate_bandwidth(&mut self, active_children: Option<u64>) -> NonZeroRatePerSecond {
+        let active_children =
+            active_children.unwrap_or_else(|| self.peers_count.load(Ordering::Acquire));
```

Comment:
While this code looks okay, according to [this](https://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering), I wonder if we really significantly benefit from using non-sequential memory ordering? Ie how many threads will call methods `calculate_bandwidth` / `request_bandwidth` / `notify_idle` and how many such calls we expect? I literally spent 30 minutes now reminding myself what memory ordering was, and I assume each next reader of this code would need to do the same. Normally, such optimizations are once reviewed and tested, wrapped up in some library code, and never touched again, so nobody needs to read complex code. Not saying this does not work, but I think we need some justification to use this optimization here.

Reply:
Changed it to `SeqCst`, which according to documentation should be exactly the same (Acquire for loads, Release for stores, Acq+Rel for fetch_modify).

Reply:
that's correct, thank you Sir!